corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py

```diff
@@ -0,0 +1,233 @@
+                                        xform_db,
+                                        dry_run,
+                                    )
+                                    f.write("Set new xmlns on submission {}\n".format(xform_instance._id))
+            for error_id in list(xform_db.error_ids) + list(app_db.error_ids):
+                f.write("Failed to save {}\n".format(error_id))
+
+    def _print_progress(self, i, total_submissions, num_fixed):
+        if i % 500 == 0 and i != 0:
+            print "Progress: {} of {} ({})".format(
+                i, total_submissions, round(i / float(total_submissions), 2)
+            )
+            print "We can fix {} of {} ({})".format(
+                num_fixed, i, round(num_fixed / float(i), 2)
+            )
+
+
+def get_submissions_without_xmlns():
+    submissions = XFormInstance.get_db().view(
+        'couchforms/by_xmlns',
+        key="undefined",
+        include_docs=False,
+        reduce=False,
+    ).all()
+    total_submissions = len(submissions)
+    submission_id_generator = (s['id'] for s in submissions)
+    submissions_doc_generator = (
+        XFormInstance.wrap(i)
+        for i in iter_docs(XFormInstance.get_db(), submission_id_generator)
+    )
+
+    total_error_submissions, error_submissions_doc_generator = _get_error_submissions_without_xmlns()
+    return (
+        total_submissions + total_error_submissions,
+        chain(submissions_doc_generator, error_submissions_doc_generator)
+    )
+
+
+def _get_error_submissions_without_xmlns():
+
+    query = (FormES()
+             .xmlns('undefined')
+             .remove_default_filter("is_xform_instance")
+             .filter(NOT(doc_type('xforminstance')))
+             .source(['_id']))
+    result = query.run()
+    total_error_submissions = result.total
+    error_submissions = (
+        XFormInstance.wrap(i)
+        for i in iter_docs(XFormInstance.get_db(), (x['_id'] for x in result.hits))
+    )
+    return total_error_submissions, error_submissions
+
+
+@quickcache(["app_id", "form.unique_id"], memoize_timeout=ONE_HOUR)
+def xforms_with_real_xmlns_possibly_exist(app_id, form):
+    """
+    Return True if there exist xforms submitted to the given app against a form
+    with the same name as the given name that have a real xmlns.
+
+    If this is True, it could indicate that the form had an xlmns, had submissions,
+    then lost the xmlns. If that's the case, we would want to give the submissions
+    this xmlns, not a new random one.
+    """
+    for form_name in form.name.values():
+        query = (FormES()
+                 .term('form.@name', form_name)
+                 .app(app_id)
+                 .filter(NOT(xmlns('undefined')))
+                 .remove_default_filter('is_xform_instance'))
+        if query.count():
+            return True
+    return False
+
+
+def set_xmlns_on_submission(xform_instance, xmlns, xform_db, dry_run):
```

Comment:
I think it might be good if this function also saved a property to the form for auditing purposes just in case we ever need to go back and figure out something in the data as a result of this migration doing something unexpected. I think that could just be a property called `form_migrated_from_undefined_xmlns` and a boolean or a date or something.